Add message-format=json support

roblabla opened the pull request:
Fixes #45 

Most of the structs come directly from cargo or rustc. This is not ready yet, there are a few things that need to be done. Most importantly, the enums (mostly Message) need to have a catch-all variant so that it won't break if new variants are added. Also, I think we'd do well to change the names of some structs here and there.

Also, I need to add some tests and fix the doctest example.

roblabla commented:
For my sanity's sake, I'm currently working on fixing serde-rs/serde#912. This will allow turning Message into

```
enum Message {
    CompilerArtifact(Artifact),
    CompilerMessage(FromCompiler),
    BuildScript(BuildScriptExecution),
    #[doc(hidden)]
    #[serde(other)]
    Unknown
}
```

Blocking on https://github.com/serde-rs/serde/pull/1382

oli-obk commented on src/diagnostic.rs:
```diff
@@ -0,0 +1,118 @@
+//! This module contains `Diagnostic` and the types/functions it uses for deserialization.
```
This file should be shared with rustfix in some manner

cc @killercup

roblabla commented:
Serde 1.0.79 is released, and with it #[serde(other)] is now available. I pushed a commit updating the serde dep to reflect that.

Question: Should all enums currently having a `DoNotMatchExhaustively` arm be switched to `#[serde(other)]`? There are only two as of this PR (not counting Message): `dependency::DependencyKind` and `diagnostic::Applicability`.

It's not a breaking change API-wise, as the arm is #[doc(hidden)]'d. Basically:

```rust
// Before
assert!(serde_json::from_str::<DependencyKind>(r#" "unknown" "#).is_error());

// After
assert_eq!(serde_json::from_str::<DependencyKind>(r#" "unknown" "#),
     Ok(DependencyKind::Unknown));
```

konstin commented:
I just tried the branch and I really like it. It makes a nice improvements over my earlier self-written structs, esp. with `parse_message_stream`.

One thing that would be useful is a `ToString` impl on `FromCompiler` using the `rendered` message. (I also wonder why `rendered` is an `Option<String>`. Is there a case were there is no rendered message?)

konstin commented on src/lib.rs:
```diff
@@ -272,6 +302,95 @@ pub enum CargoOpt {
+    /// The enabled features for this artifact
+    pub features: Vec<String>,
+    /// The full paths to the generated artifacts
+    pub filenames: Vec<String>,
```
Might those be better as PathBuf than as String?

konstin replied:
Imho workspace_root should also be changed, but maybe in a different PR. @oli-obk What do you think?

mati865 replied:
@konstin Oliver is away until October.

konstin commented on src/lib.rs:
```diff
@@ -87,6 +87,33 @@
+//!     .stdout(Stdio::piped())
+//!     .spawn().unwrap();
+//!
+//! for message in cargo_metadata::parse_message_stream(command.stdout.unwrap()) {
```
Imo this example should `.take()` before unwrapping. Otherwise you can't wait for the command to finish afterwards.

konstin commented:
What would be required to get this merged and released on crates.io? I'm using those types in pyo3-pack and would like to release on crates.io without vendoring tricks.

roblabla commented:
I'll get around to addressing all the comments in the PR tomorrow (Using PathBuf, changing enums into `#[serde(other)]`, fixing the doctest, implementing ToString). Then I think it should be ready for merging?

epage commented on src/diagnostic.rs:
```diff
@@ -0,0 +1,119 @@
+    /// The associated error code for this diagnostic
+    pub code: Option<DiagnosticCode>,
+    /// "error: internal compiler error", "error", "warning", "note", "help"
+    pub level: String,
```
Is there a reason this isn't an enum?

epage replied:
I guess the same could be said for `cargo-metadata`s existing `kind`, `crate_types`, `opt_levels`

oli-obk replied:
Yea, this was mostly chose for forward compatibility

epage commented:
I've forked this in my PR for escargot

Main differences
- I use `Cow` / `serde(borrowed)` everywhere
- I went ahead and switched to `Path`
- `Diagnostic::level` is an enum. 
- I have a `strict_unstable` feature that removes the `Unknown` variants and adds `serde(deny_unknown_fields)` that I use with my tests in the hope that it will help me identify when new fields are added

In addition, I went ahead and forked `WorkspaceMember` and `Target` rather than re-exporting `cargo-metadata`.  I find it curious that `WorkspaceMember` serializes to `raw: String` with parse functions rather than the components it makes up but I went ahead and kept that.

oli-obk commented on src/lib.rs:
```diff
@@ -339,3 +458,13 @@ pub fn metadata_run(manifest_path: Option<&Path>, deps: bool, features: Option<C
+/// An iterator of Message.
+type MessageIterator<R> = serde_json::StreamDeserializer<'static, serde_json::de::IoRead<R>, Message>;
+
+// TODO: Should I use de::Read instead, to support deserializing from a slice?
```
If it's strictly more powerful to use `de::Read`, then by all means

roblabla replied:
On investigating this a bit, de::Read isn't very convenient. It's implemented on serde_json::de::IoRead, serde_json::de::SliceRead, etc... and not on the actual libstd. Besides, I have a hard time envisioning under which situation would someone try to read an array of cargo messages from a byte array instead of an std::io::Read...

oli-obk commented on src/lib.rs:
```diff
@@ -339,3 +458,13 @@ pub fn metadata_run(manifest_path: Option<&Path>, deps: bool, features: Option<C
+
+// TODO: Should I use de::Read instead, to support deserializing from a slice?
+/// Creates an iterator of Message from a Read outputting a stream of JSON
+/// messages. For usuage information, look at the top-level documentation.
```
```suggestion
/// messages. For usage information, look at the top-level documentation.
```

roblabla commented:
Superseded by #64
